[WGSL] Add scaffolding for the type checker

tadeuzagallo opened the pull request:
#### dc9dac69fd36796e1cd101057c90b12a068ad720
<pre>
[WGSL] Add scaffolding for the type checker
<a href="https://bugs.webkit.org/show_bug.cgi?id=251524">https://bugs.webkit.org/show_bug.cgi?id=251524</a>
&lt;rdar://problem/104921530&gt;

Reviewed by Myles C. Maxfield.

Start working on the type checker. For now I added most of the structure I had
in mind, but didn&apos;t really find a clear stopping point, so there are a couple
hacks to keep the compiler happy, but I didn&apos;t want the PR to get too large.

So far, I added the core pieces for the inference algorithm, and the visitors for
all the nodes we should need to get the hello-compute shader working, but the actual
implementation will be added in a follow up PR.

* Source/WebGPU/WGSL/ContextProvider.h:
* Source/WebGPU/WGSL/ContextProviderInlines.h:
(WGSL::ContextProvider&lt;Value&gt;::Context::add):
(WGSL::ContextProvider&lt;Value&gt;::introduceVariable const):
* Source/WebGPU/WGSL/TypeCheck.cpp: Added.
(WGSL::TypeChecker::TypeChecker):
(WGSL::UnificationContext::UnificationContext):
(WGSL::UnificationContext::~UnificationContext):
(WGSL::UnificationContext::unify):
(WGSL::UnificationContext::substitute):
(WGSL::UnificationContext::solveConstraints):
(WGSL::UnificationContext::runSubstitutions):
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::variable):
(WGSL::TypeChecker::resolve):
(WGSL::TypeChecker::inferred):
(WGSL::TypeChecker::unify):
(WGSL::TypeChecker::substitute):
(WGSL::typeCheck):
* Source/WebGPU/WGSL/TypeCheck.h: Copied from Source/WebGPU/WGSL/ContextProvider.h.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259752@main">https://commits.webkit.org/259752@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eaa3132c261564f9a1332c6cdea1e7bfabfeeac6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105829 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115027 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6096 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98061 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111579 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95387 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39919 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94278 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27028 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8168 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28380 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8650 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14277 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47928 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6745 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10207 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -0,0 +1,321 @@
+private:
+    void solveConstraints()
+    {
+        // FIXME: implement this
```
Is this really so complicated that it has to be described as a graph of constraints? When we were designing WGSL we tried pretty hard to make sure that every expression had a well-defined type that was context-independent. Why can't we just naively make this a single traversal from leaves to root, annotating types at each node?

tadeuzagallo replied:
That's a good point, maybe it's unnecessary? I often find it a little easier to reason about it this way, and I was considering the combination of overload resolution with the conversion rules. I'll keep it this way for now, but I'll try to remove it once I get to implementing the actual checking.

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -0,0 +1,321 @@
+        visit(variable);
+
+    for (auto& function : m_shaderModule.functions())
+        visit(function);
```
Visiting a function doesn't visit its body? That seems confusing...

tadeuzagallo replied:
First we declare all the functions, then we visit the bodies.

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -0,0 +1,321 @@
+
+    // FIXME: allocate and build struct type from struct members
+    Type* structType = nullptr;
+    ContextProvider::introduceVariable(structure.name(), structType);
```
Is this going to work in the presence of order-independent top-level declarations?

tadeuzagallo replied:
Hum, we might have to do 2 passes (one to introduce the empty struct type and another to fill it)

litherum commented on Source/WebGPU/WGSL/TypeCheck.cpp:
```diff
@@ -0,0 +1,321 @@
+    // FIXME: allocate new type variables
+    return nullptr;
+}
+
```
I kind of think we should do the Behavior checker too while we're at it, since it composes nicely with the expression type checker (that you've got here).

The Behavior checker is the one that assigns a set of Behaviors to each statement, which correspond to the ways that control flow can exit that statement. From memory, the set of Behaviors is {Passthrough, Continue, Break, Return}. The Behavior of a conditional is the union of the conditions of both of its sides, etc. Using this kind of analysis, we can validate things like "no continue statement can appear outside of a loop" and "if a break appears in a block, it's the last statement in the block" and "if a statement has a non-void return type, all control flow paths must end in a return"

tadeuzagallo committed:
[WGSL] Add scaffolding for the type checker
https://bugs.webkit.org/show_bug.cgi?id=251524
<rdar://problem/104921530>

Reviewed by Myles C. Maxfield.

Start working on the type checker. For now I added most of the structure I had
in mind, but didn't really find a clear stopping point, so there are a couple
hacks to keep the compiler happy, but I didn't want the PR to get too large.

So far, I added the core pieces for the inference algorithm, and the visitors for
all the nodes we should need to get the hello-compute shader working, but the actual
implementation will be added in a follow up PR.

* Source/WebGPU/WGSL/ContextProvider.h:
* Source/WebGPU/WGSL/ContextProviderInlines.h:
(WGSL::ContextProvider<Value>::Context::add):
(WGSL::ContextProvider<Value>::introduceVariable const):
* Source/WebGPU/WGSL/TypeCheck.cpp: Added.
(WGSL::TypeChecker::TypeChecker):
(WGSL::UnificationContext::UnificationContext):
(WGSL::UnificationContext::~UnificationContext):
(WGSL::UnificationContext::unify):
(WGSL::UnificationContext::substitute):
(WGSL::UnificationContext::solveConstraints):
(WGSL::UnificationContext::runSubstitutions):
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::variable):
(WGSL::TypeChecker::resolve):
(WGSL::TypeChecker::inferred):
(WGSL::TypeChecker::unify):
(WGSL::TypeChecker::substitute):
(WGSL::typeCheck):
* Source/WebGPU/WGSL/TypeCheck.h: Copied from Source/WebGPU/WGSL/ContextProvider.h.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259752@main
